src/lib.rs

```diff
@@ -819,187 +835,197 @@ impl<L: SynthLanguage> Synthesizer<L> {
-                    self.egraph.number_of_classes(),
-                );
+        // run lifting rules
+        log::info!("running lifting rules");
+        let runner = self
+            .mk_cvec_less_runner(self.egraph.clone())
+            .with_iter_limit(self.params.eqsat_iter_limit + 1); // increase iter limit by 1 to ensure saturation
+        let rewrites: Vec<&Rewrite<L, SynthAnalysis>> = self.lifting_rewrites.iter().collect();
+        let (new_egraph, found_unions) = self.run_rewrites_with_unions(rewrites, runner);
 
-                chunk_num += 1;
-                for node in chunk {
-                    if iter > self.params.ema_above_iter {
-                        let rec = node.to_recexpr(|id| self.egraph[id].data.simplest.as_ref());
-                        L::add_domain_expr(&mut self, &rec);
-                    } else {
-                        L::add_domain_node(&mut self, node.clone());
+        // These unions are candidate rewrite rules.
+        // Do not apply unions on `egraph` since we are
+        // using `new_egraph` after this point
+        for ids in found_unions.values() {
+            for win in ids.windows(2) {
+                if win[0] != win[1]
+                    && self.egraph[win[0]].data.is_extractable
+                    && self.egraph[win[1]].data.is_extractable
+                {
+                    let extract = Extractor::new(&self.egraph, ExtractableAstSize);
+                    let (_, e1) = extract.find_best(win[0]);
+                    let (_, e2) = extract.find_best(win[1]);
+                    if let Some(mut eq) = Equality::new(&e1, &e2) {
+                        if e1 != e2 &&
+                            e1.as_ref().iter().all(|x| x.is_extractable()) &&        // extractable and valid
+                            e2.as_ref().iter().all(|x| x.is_extractable())
+                        {
+                            if let ValidationResult::Valid = L::validate(self, &eq.lhs, &eq.rhs) {
+                                eq.ids = Some((win[0], win[1]));
+                                if !new_candidate_eqs.contains_key(&eq.name) {
+                                    log::debug!("  Candidate {}", eq);
+                                    new_candidate_eqs.insert(eq.name.clone(), eq);
+                                }
+                            }
+                        }
                     }
                 }
+            }
+        }
 
-                // run new eqs
-                log::info!("running eqsat with {} new rules", self.new_eqs.len());
-                let mut runner = self.mk_cvec_less_runner(self.egraph.clone());
-                let rewrites: Vec<&Rewrite<L, SynthAnalysis>> =
-                    self.new_eqs.values().flat_map(|eq| &eq.rewrites).collect();
-                runner = runner.run(rewrites);
-                self.egraph = runner.egraph;
-                self.egraph.rebuild();
+        self.egraph = new_egraph;
```

Comment:
CONFUSION!! why do we do this again?

Reply:
Lifting => the runner egraph is a denoted version, we want to keep the extra nodes